source/game/kart/CollisionGroup.cc

```diff
@@ -0,0 +1,205 @@
+#include "CollisionGroup.hh"
+
+#include <test/TestDirector.hh>
+
+namespace Kart {
+
+void CollisionData::reset() {
+    tangentOff.setZero();
+    floorNrm.setZero();
+    vel.setZero();
+    relPos.setZero();
+    movement.setZero();
+    speedFactor = 1.0f;
+    rotFactor = 0.0f;
+    closestFloorFlags = 0;
+    closestFloorSettings = 0xffffffff;
+    intensity = 0.0f;
+
+    floor = false;
+}
+
+Hitbox::Hitbox() : m_bspHitbox(nullptr) {}
+
+Hitbox::~Hitbox() = default;
+
+void Hitbox::update(f32 param_1, f32 totalScale, const EGG::Vector3f &scale, const EGG::Quatf &rot,
+        const EGG::Vector3f &pos) {
+    f32 fVar1 = 0.0f;
+    if (scale.y < param_1) {
+        fVar1 = (param_1 - scale.y) * m_bspHitbox->radius;
+    }
+
+    EGG::Vector3f scaledPos = m_bspHitbox->position * scale;
+    scaledPos.y = (m_bspHitbox->position.y + totalScale) * scale.y + fVar1;
+
+    m_relPos = rot.rotateVector(scaledPos);
+    m_worldPos = m_relPos + pos;
+}
+
+void Hitbox::reset() {
+    m_worldPos.setZero();
+    m_lastPos.setZero();
+    m_relPos.setZero();
+}
+
+void Hitbox::setRadius(f32 radius) {
+    m_radius = radius;
+}
+
+void Hitbox::setBspHitbox(const BSP::Hitbox *hitbox) {
+    m_bspHitbox = hitbox;
+}
+
+void Hitbox::setWorldPos(const EGG::Vector3f &pos) {
+    m_worldPos = pos;
+}
+
+void Hitbox::setLastPos(const EGG::Vector3f &pos) {
+    m_lastPos = pos;
+}
+
+void Hitbox::setLastPos(const EGG::Vector3f &scale, const EGG::Matrix34f &pose) {
+    f32 yScaleFactor = scale.y;
+    EGG::Vector3f scaledPos = m_bspHitbox->position;
+    scaledPos.x *= scale.x;
+    scaledPos.z *= scale.z;
+
+    if (scale.y != scale.z && scale.y < 1.0f) {
+        scaledPos.y += (1.0f - scale.y) * m_radius;
+        yScaleFactor = scale.z;
+    }
+
+    scaledPos.y *= yScaleFactor;
+    m_lastPos = pose.ps_multVector(scaledPos);
+}
+
+const BSP::Hitbox *Hitbox::bspHitbox() const {
+    return m_bspHitbox;
+}
+
+EGG::Vector3f Hitbox::worldPos() const {
+    return m_worldPos;
+}
+
+EGG::Vector3f Hitbox::lastPos() const {
+    return m_lastPos;
+}
+
+f32 Hitbox::radius() const {
+    return m_radius;
+}
+
+CollisionGroup::CollisionGroup() : m_bspHitboxCount(0), m_hitboxScale(1.0f) {
+    m_collisionData.reset();
+}
+
+CollisionGroup::~CollisionGroup() {
+    delete[] m_hitboxes->data();
+    delete m_hitboxes;
+}
+
+f32 CollisionGroup::initHitboxes(const std::array<BSP::Hitbox, 16> &hitboxes) {
+    u16 bspHitboxCount = 0;
+
+    for (const auto &hitbox : hitboxes) {
+        if (parse<u16>(hitbox.enable)) {
+            ++bspHitboxCount;
+        }
+    }
+
+    m_bspHitboxCount = bspHitboxCount;
+
+    m_hitboxes = new std::span<Hitbox>(new Hitbox[m_bspHitboxCount], m_bspHitboxCount);
+
+    for (u16 hitboxIdx = 0; hitboxIdx < m_hitboxes->size(); ++hitboxIdx) {
+        (*m_hitboxes)[hitboxIdx].setBspHitbox(&hitboxes[hitboxIdx]);
+    }
```

Comment:
We need to have a larger discussion around whether we want to iterate based on length-related variables like `m_bspHitboxCount` or if we want to shift towards solely using `std::vector::size()`. If the latter, then we can probably discard `m_bspHitboxCount` in its entirety.